messaging/broker.go

```diff
@@ -332,67 +348,249 @@ func (b *Broker) applyUnsubscribe(m *Message) {
 // This is implemented as a separate type because it is not meant to be exported.
 type brokerFSM Broker
 
-// Apply executes a raft log entry against the broker.
-func (fsm *brokerFSM) Apply(e *raft.LogEntry) error {
+// MustApply executes a raft log entry against the broker.
+// Non-repeatable errors such as system or disk errors must panic.
+func (fsm *brokerFSM) MustApply(e *raft.LogEntry) {
 	b := (*Broker)(fsm)
 
-	// Ignore internal raft entries.
-	if e.Type != raft.LogEntryCommand {
-		// TODO: Save index.
-		return nil
-	}
+	// Save highest applied index.
+	// TODO: Persist to disk for raft commands.
+	b.index = e.Index
 
-	// Decode the message from the raft log.
+	// Create a message with the same index as Raft.
 	m := &Message{}
-	err := m.UnmarshalBinary(e.Data)
-	assert(err == nil, "message unmarshal: %s", err)
 
-	// Add the raft index to the message.
-	m.Index = e.Index
-
-	// Update the broker configuration.
-	switch m.Type {
-	case CreateReplicaMessageType:
-		b.applyCreateReplica(m)
-	case DeleteReplicaMessageType:
-		b.applyDeleteReplica(m)
-	case SubscribeMessageType:
-		b.applySubscribe(m)
-	case UnsubscribeMessageType:
-		b.applyUnsubscribe(m)
+	// Decode commands into messages.
+	// Convert internal raft entries to no-ops to move the index forward.
+	if e.Type == raft.LogEntryCommand {
+		// Decode the message from the raft log.
+		err := m.UnmarshalBinary(e.Data)
+		assert(err == nil, "message unmarshal: %s", err)
+
+		// Update the broker configuration.
+		switch m.Type {
+		case CreateReplicaMessageType:
+			b.mustApplyCreateReplica(m)
+		case DeleteReplicaMessageType:
+			b.mustApplyDeleteReplica(m)
+		case SubscribeMessageType:
+			b.mustApplySubscribe(m)
+		case UnsubscribeMessageType:
+			b.mustApplyUnsubscribe(m)
+		}
+	} else {
+		// Internal raft commands should be broadcast out as no-ops.
+		m.TopicID = BroadcastTopicID
+		m.Type = InternalMessageType
 	}
 
+	// Set the raft index.
+	m.Index = e.Index
+
 	// Write to the topic.
 	t := b.createTopicIfNotExists(m.TopicID)
 	if err := t.encode(m); err != nil {
-		return fmt.Errorf("encode: %s", err)
+		panic("encode: " + err.Error())
 	}
-
-	return nil
 }
 
 // Index returns the highest index that the broker has seen.
 func (fsm *brokerFSM) Index() (uint64, error) {
-	// TODO: Retrieve index.
-	return 0, nil
+	b := (*Broker)(fsm)
+	return b.index, nil
 }
 
 // Snapshot streams the current state of the broker and returns the index.
 func (fsm *brokerFSM) Snapshot(w io.Writer) (uint64, error) {
+	b := (*Broker)(fsm)
+
 	// TODO: Prevent truncation during snapshot.
-	// TODO: Lock and calculate header.
-	// TODO: Retrieve snapshot index.
-	// TODO: Stream each topic.
-	return 0, nil
+
+	// Calculate header under lock.
+	b.mu.RLock()
+	s, err := fsm.createSnapshot()
+	b.mu.RUnlock()
+	if err != nil {
+		return 0, fmt.Errorf("create snapshot: %s", err)
+	}
+
+	// Encode snapshot header.
+	buf, err := json.Marshal(&s)
```

Comment:
I don't understand how this only encodes the "snapshot header" since the code passes `s` -- a full snapshot -- to the `Marshal` function. (Perhaps it will become clear as I continue through the code.)